scientific-computing-with-python/learn-inheritance-by-building-an-equation-solver/quadratic_equation_solver.py

```diff
@@ -0,0 +1,186 @@
+            self.results.add('No real roots')
+            return roots
+                
+        x1 = (- self.b + (self.delta)**0.5)/(2 * self.a)
+        x2 = (- self.b - (self.delta)**0.5)/(2 * self.a)        
+        if self.delta == 0:        
+            roots = {'x': x1}
+            self.results.add(f'x = {round(x1, 3)}')
+            return roots
+                       
+        roots = {'x1': x1, 'x2': x2}
+        self.results.add(f'x1 = {round(x1, 3)}')
+        self.results.add(f'x2 = {round(x2, 3)}')
+        return roots
+    
+    def analyze(self):                
+        x = - self.b / (2 * self.a)
+        y = self.a * x**2 + self.b * x + self.c
+        vertex = {'x': x, 'y': y}                    
+        if self.a > 0:
+            concavity = 'upwards'
+            vertex['m'] = 'min'            
+        else:
+            concavity = 'downwards'
+            vertex['m'] = 'max'
+        coord = f'{(round(x, 3), round(x, 3))}'
+        self.details.add(f'concavity = {concavity:>12}\n{vertex["m"]} = {coord:>18}')
+        return vertex, concavity
+    
+
+class QuadraticInequality(QuadraticEquation):
+    type = 'Inequality'
+
+    def __init__(self, sign='>', a=0, b=0, c=0,):
+        super().__init__(a, b, c)        
+        self.sign = sign
+        self.line = self.line.replace('=', self.sign )
+
+    def __invert_sign(self):
+        if self.a < 0:
+            self.a *= -1
+            self.b *= -1
+            self.c *= -1
+            if '>' in self.sign:
+                self.sign = self.sign.replace('>', '<')
+            elif '<' in self.sign:
+                self.sign = self.sign.replace('<', '>')
+
+    def solve(self):
+        self.__invert_sign()
+        if self.delta < 0:
+            if '>' in self.sign:
+                self.results.add('∀ x')
+            elif '<' in self.sign:
+                self.results.add('∄ x')
+            return self.results
+        
+        x1 = (- self.b + (self.delta)**0.5)/(2 * self.a)
+        x2 = (- self.b - (self.delta)**0.5)/(2 * self.a)        
+        if self.delta == 0:
+            if self.sign == '>':
+                self.results.add(f'x ≠ {round(x1, 3)}')
+            elif self.sign == '>=':
+                self.results.add(f'∀ x')
+            elif self.sign == '<':
+                self.results.add(f'∄ x')
+            elif self.sign == '<=':
+                self.results.add(f'x = {round(x1, 3)}')
+            return self.results
+        
+        x_min = min(x1, x2)
+        x_max = max(x1, x2)
+        e = '=' if '=' in self.sign else ''
+        if '>' in self.sign:            
+            self.results.update((f'x <{e} {round(x_min, 3)}', f'x >{e} {round(x_max, 3)}'))
+        elif '<' in self.sign:            
+            self.results = {f'{round(x_min, 3)} <{e} x <{e} {round(x_max, 3)}'}        
+        return self.results
+
+    def analyze(self):
+        raise AttributeError("'QuadraticInequality' object has no attribute 'analyze'")
```

Comment:
To be honest there isn't that much inheritance between these two classes. Line above is a good example for that. One of the _SOLID_ principle is that it should be possible to replace the instance of parent with instance of child, without changing behavior. In other words child class should be doing everything (and possibly more) as the parent class in the same way. `analyze` would yell at somebody expecting that :slightly_smiling_face: 

It's not easy to come up with good inheritance example, but I think the the equations idea can still be used, with slight reorganizing and tackling it from different angle. I'd propose to introduce abstract base class `Equation` (I guess it doesn't necessarily needs to be abstract, but this is a great place for it) which will be a parent class of equations.

```python
from abc import ABC, abstractmethod
class Equation(ABC):
    """"""
    @abstractmethod
    def solve(self):
        pass

    @abstractmethod
    def details(self):
        pass

    @abstractmethod
    def coefficients(self):
        pass
```
Because of being abstract parent, it will make sure the child classes are implementing methods that are required for class to work with the _Solver_:

```python
class Solver:
    def __init__(self, equation):
        self.equation = equation
    def create_output(self):
        ...
    # other methods presenting/preparing result
```
`Solver` isn't a great name for this class, because the `equation` effectively solves itself on it own. I don't have anything better for it right now.

Then we have concrete equations, that must implement abstract methods from `Equation`. They can of course add more methods that apply to the equation type that they are, but the _core interface_ with which `Solver` uses them is defined in the abstract base class `Equation`:
```python
class QuadraticEquation(Equation):
    ....

class QuadraticInequality(Equation):
    ....
```
But also there can be simpler (or harder) concrete equations defined, if somebody (camper) wants and is able to create concrete class, which still will work with the _Solver_.
```python
class LinearEquation(Equation):
    ....
```

Reply:
I agree, I was struggling to find a good way to teach something useful and meaningful at the same time.

Probably, this is the right place to introduce abstract classes and it can improve the overall project. I'll try to restructure the prototype.